### ISSUE TITLE
Parallelize recover_dynamics

### ISSUE THREAD

**Issue by WeilerP** [theislab/scvelo#342]
## Description

To boost performance, the function `recover_dynamics` can be parallelized.

**Comment by yannikschaelte** [comment]
Sounds good to me!

**Comment by yannikschaelte** [comment]
Ok, had a short look. I like that the use of joblib simplifies a few things. Some differences I noted in #265 to maybe keep in mind, nothing critical:

* If I see correctly, each worker returns the full DynamicsRecovery objects. Therefore, all internal variables (all matrices) would be copied `n_jobs` times. For 8 cores ok, but may lead to memory problems on more cores. This problems also exists in #331 potentially and can be overcome by having the workers really only return the output estimates of interest. This should be feasible, the dms are currently only returned due to some legacy part further down where the last dm is used.
*  The batches given to each worker are pre-defined. Therefore, if e.g. all the costly genes (observed some heterogeneity) where optimization takes longer happen to be in one batch, it cannot dynamically adjust and that process will take longer than the others. This is why I used those "mini-batches". The additional queuing communication may however have lead to a different overhead (use larger mini-batches?).
* I think for large-scale problems (runtime >1day), the goal, as a second step, should not be multiprocessing on a laptop pc, but distributed computation on a cluster. For this, a parallelization using e.g. dask, redis, spark, mpi would be preferable. This is easily possible in #331, in #265 it should also be possible via joblib's custom backend (which is in test mode still?), but I have not checked that in detail.
* For `n_jobs = 1`, there may be a relatively small overhead from unnecessary serialization.

**Comment by WeilerP** [comment]
Yes, I agree. Locally, running on 12 cores, I did not see a major difference concerning the memory usage between the two approaches. Granted, I did not perform a in-depth analysis but only monitored the process via `htop`.

We will also refactor and restructure the scVelo codebase in the upcoming weeks. For example, the class `DynamicsRecovery` can be rewritten to not have the entire `AnnData` as an attribute but only the minimum number of data passed to it. After having implemented such changes, we can reiterate on the implemented parallelisation scheme.
